### PR TITLE
DAOS-7821 flake: Add flake8 config file and resolve some warnings.

### PR DESCRIPTION
**Pull request by ashleypittman**
Resolve all instances of some classes of warnings and enable these
in GHA.
Write a config file for local use.
Add a script as a basis for a git-hook to run checks pre-commit.

Skip-build: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: utils/githooks/pre-commit.d/20-flake.sh
Comment (brianjmurrell) [inline utils/githooks/pre-commit.d/20-flake.sh:22]:
```suggestion
if ! git diff -u | flake8 --diff; then
    exit ${PIPESTATUS[@]}
```

or even just let `set -e` do the work for you:

```suggestion
git diff -u | flake8 --diff
```

should have the same result IIRC.

File: utils/githooks/pre-commit.d/20-flake.sh
Comment (brianjmurrell) [inline utils/githooks/pre-commit.d/20-flake.sh:24]: Should this be error checked with an `if/then/fi` wrapper?

File: utils/githooks/pre-commit.d/20-flake.sh
Comment (brianjmurrell) [inline utils/githooks/pre-commit.d/20-flake.sh:38]: Maybe a warning here about the `master` assumption?

**Review by brianjmurrell — CHANGES_REQUESTED** [review]
This of course cannot land until (and if, even) #8155 lands as it depends on the structure that that PR adds.

File: .github/workflows/flake.yml
Comment (ashleypittman) [reply .github/workflows/flake.yml:23]: I seem to recall when I added this I set it to "latest" and you had me hard-code a version.

File: utils/githooks/pre-commit.d/20-flake.sh
Comment (ashleypittman) [reply utils/githooks/pre-commit.d/20-flake.sh:38]: I thought I'd updated this not to assume master.  Let me check.

**Comment by ashleypittman** [comment]
> This of course cannot land until (and if, even) #8155 lands as it depends on the structure that that PR adds.

It should be fine to land either before or after #8155, although as a commit-hook it won't be enabled unless they're both present obviously.

File: .github/workflows/flake.yml
Comment (ashleypittman) [reply .github/workflows/flake.yml:23]: Apologies, you asked for this before https://github.com/daos-stack/daos/pull/8152#discussion_r804784669

File: utils/githooks/pre-commit.d/20-flake.sh
Comment (brianjmurrell) [inline utils/githooks/pre-commit.d/20-flake.sh:39]: Yeah, like ShellCheck is advising:

```suggestion
	if ! TARGET=$(gh pr view "$BRANCH" --json baseRefName -t "{{.baseRefName}}")
```

File: utils/githooks/pre-commit.d/20-flake.sh
Comment (ashleypittman) [reply utils/githooks/pre-commit.d/20-flake.sh:43]: Had it in my local copy, although it was also incorrect.

File: .github/workflows/flake.yml
Comment (ashleypittman) [inline .github/workflows/flake.yml:23]: This is now using 3.10.2 and as you say should auto-update.

File: utils/githooks/pre-commit.d/20-flake.sh
Comment (ashleypittman) [reply utils/githooks/pre-commit.d/20-flake.sh:2]: I don't think the "command" tests will work with -e set, and this all displays to the terminal when you commit so whilst -x is useful to debug the script we don't want it enabled for landing.

File: utils/githooks/pre-commit.d/20-flake.sh
Comment (ashleypittman) [reply utils/githooks/pre-commit.d/20-flake.sh:24]: Done.

File: utils/githooks/pre-commit.d/20-flake.sh
Comment (brianjmurrell) [inline utils/githooks/pre-commit.d/20-flake.sh:49]:
```suggestion
    echo "Checking against branch ${TARGET}"
```

File: utils/githooks/pre-commit.d/20-flake.sh
Comment (brianjmurrell) [reply utils/githooks/pre-commit.d/20-flake.sh:2]: All of your `command` uses are wrapped with an `if/then/fi` so you can use `set -e` here.

Comment (ashleypittman) [reply utils/githooks/pre-commit.d/20-flake.sh:2]: You learn something new every day, preparing a new version now.

File: src/tests/ftest/soak/utils.py
Comment (daltonbohning) [inline src/tests/ftest/soak/utils.py:1161]: Is this really our preferred style? I don't like this

File: utils/fix_cmocka_xml.py
Comment (daltonbohning) [inline utils/fix_cmocka_xml.py:19]: Does `path+fn` not need spaces?

File: src/tests/ftest/soak/utils.py
Comment (ashleypittman) [reply src/tests/ftest/soak/utils.py:1161]: Flake is checking that tha ] is aligned just to the right of the ], I suspect it would also allow it if the ] was at the end of the previous line however.

File: utils/fix_cmocka_xml.py
Comment (ashleypittman) [reply utils/fix_cmocka_xml.py:19]: doesn't seem to, I hadn't really read this code properly and looking now it's not great is it.

File: src/tests/ftest/soak/utils.py
Comment (daltonbohning) [reply src/tests/ftest/soak/utils.py:1161]: I more so meant that I prefer:
```
    prepend_cmds = [
        "set -e",
        "echo Job_Start_Time `date \\+\"%Y-%m-%d %T\"`",
        "daos pool query {} ".format(self.pool[1].uuid),
        "daos pool query {} ".format(self.pool[0].uuid)]
```
OR
```
    prepend_cmds = [
        "set -e",
        "echo Job_Start_Time `date \\+\"%Y-%m-%d %T\"`",
        "daos pool query {} ".format(self.pool[1].uuid),
        "daos pool query {} ".format(self.pool[0].uuid)
    ]
```

I despise having to indent something that isn't aligned with tab width. But we do that in all of our Python and C code, so I suspect I've already lost that battle. To my point: think about what happens if the length of the variable name `prepend_cmds` changes. You have to reformat the whole list! It's also a silly notion that indentation is tied to the length of a variable name or function call. I digress

File: utils/fix_cmocka_xml.py
Comment (daltonbohning) [reply utils/fix_cmocka_xml.py:19]: I decided to not comment on the implementation, just the style :)

File: src/tests/ftest/soak/utils.py
Comment (ashleypittman) [reply src/tests/ftest/soak/utils.py:1161]: The later is what we had before (which I personally don't like) but I think the former would work.  The whole flake checking is new, what I did for this PR was ran "flake8 --statistics" and fixed/resolved any errors that had fewer than 10 or so warnings and enabled the check.  I'll try the former from above, and if that doesn't work then I'll back out this part and leave that check disabled.

> To my point: think about what happens if the length of the variable name prepend_cmds changes. You have to reformat the whole list! It's also a silly notion that indentation is tied to the length of a variable name or function call. I digress
The point of having auto-formatting is that doesn't matter any more, yes the code gets re-indented but it happens automatically, and if everybody does it then whilst everybody does it the resulting code is the same so there are no merge conflicts and the code having a consistent style really does make it easier to read.

Comment (daltonbohning) [reply src/tests/ftest/soak/utils.py]: In the end, I think consistency is most important.

File: utils/githooks/pre-commit.d/20-flake.sh
Comment (brianjmurrell) [reply utils/githooks/pre-commit.d/20-flake.sh:23]: Still seems like this would be useful to wrap with an `if` to not confuse the user into thinking they have python code errors if/when this call fails:

```suggestion
if ! BRANCH=$(git rev-parse --abbrev-ref HEAD > /dev/null 2>&1); then
    echo "Failed to determine branch with git rev-parse"
    exit 1
fi
```

**Review by brianjmurrell — DISMISSED** [review]

File: utils/githooks/pre-commit.d/20-flake.sh
Comment (ashleypittman) [reply utils/githooks/pre-commit.d/20-flake.sh]: Done.  Githut automatically marked this as resolved when I accepted your suggestion.

Comment (brianjmurrell) [reply utils/githooks/pre-commit.d/20-flake.sh]: > Githut automatically marked this as resolved when I accepted your suggestion.

Interesting.  Doesn't seem unreasonable.  Ideally though, I think github should distinguish between it resolving a conversation because a suggestion was accepted and the PR owner resolving conversations.  But frankly, I think github should allow a project (or maybe it's a per-user preference) to disable allowing anyone except the conversation owner resolving conversations.

**Review by daltonbohning — APPROVED** [review]

**Review by brianjmurrell — APPROVED** [review]

**Pull request merged**
